### ISSUE TITLE
Language support module release cycles

### ISSUE THREAD

**Issue by pvlugter** [cloudstateio/cloudstate#289]
It's not clear when and how language support modules are released, and which versions of the cloudstate proxy and protocol these are compatible with. Especially for the java-support and node-support modules in the main repo, which are currently released as needed.

In the long-term, we should be looking to have these modules in separate repos like other language support — with their own versions, automated releases, and docs (see #104), and protocol versioning built-in (see #288).

**Comment by viktorklang** [comment]
Agreed. And we also need to rework the TCK so that the different language-supports can run them as a part of *their* build—ideally spitting out a summary of all the relevant meta data and which Cloudstate Statemodels they support.

**Comment by sleipnir** [comment]
I think that the only modules that are not separate today are Java and Node modules, all the others have completely independent life cycles. What I would like to do is to establish a versioning rule that respects the versioning of the protocol itself and that serves all language support. What I mean is that if the protocol (protos, operator, and proxy) are in version 0.5.0 and the development of Dart or Kotlin is ahead of that version (0.5.4) they should inform the version of the protocol in versioning (0.5.4_v0.5.0) or something like that because it is clear to the end user what to expect from the implementation.

@viktorklang 
I agree about the TCK but I think the compatibility matrix can be done initially simply by documenting what each language supports in its own repository, then we automate that. We need a TCK independent from the main repository for yesterday, we cannot delay with that or we will be in the dark for much longer than we should.

**Comment by viktorklang** [comment]
@sleipnir @pvlugter Given how the TCK works I think it should be rather straightforward to create a Docker image for it and have the user run it against that Docker image passing in the right parameters to it. @pvlugter, is this something you can take the lead on? Let me know how I can help.

**Comment by pvlugter** [comment]
It looks like the TCK starts other processes or Docker containers. Does it work to have it run in a Docker container itself? Simplest given what I see currently, could be to publish the TCK and each language support has its own TCK build — simply break apart what's there currently, each support has a small sbt build for the TCK and paradox.

**Comment by pvlugter** [comment]
On versioning, I think we'll expect the cloudstate protocol, proxy, and language supports to have different rhythms — so not sure if they should be aligned on version series (like 0.5.x), although maybe they are at first.

So I'd expect that at some point the protocol becomes quite stable, the proxy keeps evolving based on features that are separate from the protocol (around how akka cluster is used, autoscaling, and so on), and the language support versions are based mainly on their user APIs and how those evolve. The proxy and language supports should be clear about which protocol versions they support — maybe that's by having aligned version series, but that could be too restrictive?

**Comment by sleipnir** [comment]
> On versioning, I think we'll expect the cloudstate protocol, proxy, and language supports to have different rhythms — so not sure if they should be aligned on version series (like 0.5.x), although maybe they are at first.
> 
> So I'd expect that at some point the protocol becomes quite stable, the proxy keeps evolving based on features that are separate from the protocol (around how akka cluster is used, autoscaling, and so on), and the language support versions are based mainly on their user APIs and how those evolve. The proxy and language supports should be clear about which protocol versions they support — maybe that's by having aligned version series, but that could be too restrictive?

Ok. I agree

> It looks like the TCK starts other processes or Docker containers. Does it work to have it run in a Docker container itself? Simplest given what I see currently, could be to publish the TCK and each language support has its own TCK build — simply break apart what's there currently, each support has a small sbt build for the TCK and paradox.

I can understand the argument but I don't see the need to have tck and sbt's dependence on other projects.
The question of the execution of docker commands by the current tck happens due to the way it was built and not because of the support languages. If we build a TCK image, the support languages will simply execute commands from their own build tools that do not depend on the docker (make run, dotnet run, dart run, mvn run, and etc ...) because, they would be incorporating the tck in your own build. It is enough that the TCK image accepts input arguments or environment variables for everything to work accordingly.

**Comment by pvlugter** [comment]
Sure, we can look at having a standalone TCK docker image that runs the tests and leaves it up to the support module to start everything else.

But we'll probably still require a JVM and sbt for working with paradox.

**Comment by sleipnir** [comment]
Yes and No. we can still use the docker and still pass parameters to the TCK.
Do you say paradox because of the question of separating the documentation in the language support modules?

**Comment by pvlugter** [comment]
Yes, paradox for allowing each language support module to own its docs.

**Comment by sleipnir** [comment]
It is so sbt lives :D

**Comment by marcellanz** [comment]
@viktorklang wrote:
> ideally spitting out a summary of all the relevant meta data and which Cloudstate Statemodels they support.

@viktorklang, @sleipnir @pvlugter For the nigthly tck run I currently parse the output from each TCK run for failures and succeeding tests of the TCK and take it over to a small struct in JSON together with other data for each run and push it to a gist repository. 
I think it would be nice the TCK would spit out its data in a structured form, at least one to be parsed in a safer way like a "JSON line". This way, we also can report which state models are supported, and then which tests for each model pass or not. We could use the same data collected, for releases of the core project and the support libraries, to have a matrix to show what state models or which features are supported. It could be the source of truth and be validated solely by the TCK, which would be a functional proof of it. wdyt?

**Comment by sleipnir** [comment]
I think we would only gain from this \0/

**Comment by viktorklang** [comment]
@marcellanz Having a structured output would be really nice—perhaps even the ability to directly generate a report.

**Comment by marcellanz** [comment]
I can take that as an action item for the TCK, as I currently learn how it works, to have the output the TCK produces beeing added a structured output of its results; or anyone who likes to have a stab at it.

**technical note**: to keep things simple, it might be too much to fiddle with stderr/stdout channels for that or even having an output file produced. The TCKs image output, with now #216 supporting it, through /dev/stdout could be the channel to transport that, I'd be fine to have it the last line (JSON line) be the structured output to be parsed by a caller and having "normal" output for humans above this last line)

@viktorklang regarding having a report, I though about to have the _TCKs report_ being that structured output and for a human that runs the TCK image having a textual CLI representation of it.

Any succeeding task, like a TCK board runner, can then parse the structured output and produce whatever he likes to, like:
- a visual TKC board
- core/proxy <=> Language Support compatibility table/matrix for documentation (paradox).

We might send the textual CLI report to anyone interested or having it on a travis/CI log displayed. For "local" development, the report is just displayed for a developer to have it run locally.

So basically whenever I run a TCK image as a human, I get the results visually as CLI text and the last line is structured to be parsed by anyone interested.

Regarding a release cycle, if we release a version of the proxy, a TCK report would be the input of the compatibility matrix for what we support, for any support library covered by the TCK. For sure, user libraries documentation have to match that and we probably would not release if one of the major support libraries fail the TCK. But having that automated liberates us to having that tracked and documented manually.

**Comment by marcellanz** [comment]
With protocol versioning #288, I'm curious how support libraries should handle unsupported versions of the Cloudstate protocol and at which level we discover that we have a version clash.

I'd favour to have the proxy or main project and support libraries having diverged version numbers. As @viktorklang mentioned, we have the proxy reporting its supported major/minor version of the protocol to a user support library during entity discovery. A user support library can reject to return any `EntitySpec` if it does not support the protocol version the proxy supports. As the proxy is a side-car to the user function, this would actually be a rare case to happen I think. It's not like the user function to run on an IoT device somewhere in the ocean and a proxy discovers the entity providing it an incompatible version? But if it does, the user support library should reject to return any `EntitySpec` being returned. 

As said in #288, we might define that in the spec. Define what a semantic version means to the proxy at all. The user function might, at runtime, report that it would not return an `EntitySpec` for the proxy with a certain protocol revision or even error on that. The proxy could do nothing against that and its deployment is faulty already perhaps, because of the side-car relation between a proxy and its user function. 

@pvlugter wrote in #288: "Ideally both sides can support multiple versions and negotiate on the latest supported", this would lead to that an Akka Cluster can handle multiple versioned proxies? (I assume thats the case) Saying this, what would the story be of a whole lot of user functions, with their proxies, floating around and having different versions of the protocol supported?

**Comment by sleipnir** [comment]
@marcellanz @viktorklang @pvlugter I think the possibility of running sidecars in different versions would be nice, but I don't know if this is technically feasible, only if the protocol versions are incompatible and not the proxy itself that would be valid, because the proxy still needs to be compatible with AKKA Cluster to be viable, so perhaps the user function should inform the 'Range' of proxy versions it supports in this way the sidecar can be chosen from a bed of options

**Comment by sleipnir** [comment]
it is also worth remembering that support libraries take time to be fully compatible with a new version of Cloudstate simply because it takes effort and hours of development that are not always on contributors' priorities for a given time. So I think that a new version should not be prevented from being released because some library does not pass the TCK. The compatibility matrix should serve only as more information for users to decide what they want to start with and how.

**Comment by sleipnir** [comment]
Today i realized that the possibility of having different versions of the proxy running alongside AKKA Cluster would only be possible if the discovery step (EntityDiscovery -> ServiceInfo) was performed by Cloudstate Kubernetes Operator instead of the Proxy because it is he responsible for injecting the sidecar to the k8's Pod. Which would be fantastic if he did it, in which case if a user function declares something incorrect the entire Deployment would fail which would facilitate debugging what could go wrong (instead of looking at the logs of two we would only have to look at the description of the deployment).

**Comment by marcellanz** [comment]
interesting point @sleipnir

**Comment by marcellanz** [comment]
@sleipnir actually, I'm curious if we ever will have a user function not having the proxy _near_ its side by, but remote.

**Comment by sleipnir** [comment]
@marcellanz I don't know if it makes sense not to have the proxy. I mean, for some frameworks like Camel-K it makes perfect sense and it's fantastic, but camel doesn't have the need for a cluster to make its magic work. I think that as we have a very strong dependency on AKKA Persitence, it would not make much sense for the absence of a sidecar as the operator would have to form a cluster in the same way. Of course, this is also a workable constraint because the Operator could deploy the nodes as DaemonSets and the AKKA cluster would be the same size as the underlying Kubernetes cluster. Another approach would be a mix, user functions that do not require persistence would not require a sidecar. But this would delegate much more work to the Operator and add greater latency as well, since without sidecar the communication would be remote via the network instead of local via "ipc" where the kernel can make many optimizations as in the case of the sidecar.
Don't know is something interesting to discuss the pros and cons.
I think that this implementation of the Cloudstate protocol based on Scala + AKKA may not make sense. Perhaps other implementations could bring other solutions that would allow use without a sidecar. I really don't know, I think a 100% approach without a sidecar doesn't make a lot of sense. But it is an interesting question.
Comment there @viktorklang @pvlugter @jroper @jboner

**Comment by marcellanz** [comment]
@sleipnir I did not meant to eliminate the proxy; I thought about to have the user function running remotely from it, like in an embedded system, a legacy system.

**Comment by sleipnir** [comment]
Did you mean running on the Edge side.
Yeah! I like it.
I talk a lot with viktor about adding other Eventing middleware, maybe an AMQP, MQTT that maybe would leverage the use of Cloudstate on the Edge side. Of course, I think nothing will stop us from running the current state of Cloudstate on a [K3S](https://k3s.io/) for example.

**Comment by marcellanz** [comment]
Think about a sensor, a car. An old and clumsy legacy system written in Ruby or Python or C++. Just an idea.

**Comment by marcellanz** [comment]
I understand fully the ambition of serverless and Cloudstate, but my point is, that many (many) systems will run for a long time not on this wave of serverless-happyness and are sceptical to migrate away from it or are unable to do so by dragging a whole lot of infrastructure or dependencies out of that into this from their perspective, complete paradigm shift.

I've seen a panel of serverless experts on the Microsoft Build conference the other week and the host asked about which concepts or tools to use to help on-premise or legacy systems bringing customer into the "serverless" world... a looooong pause followed this question with stunning faces not having an answer or even a clue. This is not their fault, living in this world is perhaps very captivating. But, many many traditional companies, where file-transfers and SOAP is still "current technology", are struggling to move. There should be a story for that or a migration path. Perhaps this is the wrong project to do that. I had more than once the thought to have a user function, with its user function support embedded or placed in the right spot to cut out a certain functionality or service.

**Comment by sleipnir** [comment]
I agree.
I think that to get to this in Cloudstate one of the ways would be that we should attack the proxy deeply so that it is smart enough, robust enough, and above all, it offers the right set of tools to enable the simplification of user functions. That way I wouldn't scare the developer who is trying to live with his legacy technology because, he would look and say: - "Wow, I don't need to bring my infrastructure together for this to work."
Another way is to use tools, we must offer more tools (build, scaffolds, easier provisioning, etc.). The latter is the reason why I started Cloudstate CLI because, my goal was to abstract the complexities of kubernetes (Kubernetes is complex, scares most people) and Cloudstate itself to facilitate the entry of new users in the technology.
But it is clear that the discussion is more ambiguous than that. Only thoughts.

**Comment by marcellanz** [comment]
>he would look and say: - "Wow, I don't need to bring my infrastructure together for this to work."

Gotcha. The same we discussed once to have the user functions not interfacing databases directly but the proxy brings the database to the function through an easy to understand interface (gRPC/HTTP) and a variety of helpful distributed systems concepts with the state models Cloudstate supports.

**Comment by sleipnir** [comment]
And I think that is the point for many other things, like HTTP, Messaging. Things that are boilerplate for many projects but that make any kind of migration difficult. The same reasoning applied to the "State model" can be used for these other things.
